### ISSUE TITLE
Newly created image files in current directory not thumbnailed

### ISSUE THREAD

**Issue by 0ion9** [IgnorantGuru/spacefm#591]
Particularly after cut/paste[1], some image files may retain the generic 'picture' icon. There doesn't appear to be any special characteristic to the files that are not being thumbnailed.

Example screenshot:
![screenshot_545](https://cloud.githubusercontent.com/assets/115782/10750527/5a335ba6-7cc6-11e5-9a8c-e3e436e2383a.png)

The files `_plants-*` had thumbnails before they were cut+pasted from the right panel. The `0_*brick-wall.png` file was saved after that cut+paste was performed. (I'm happy to upload these files if need be, but I'm confident they are completely ordinary)

Restarting SpaceFM usually causes the thumbnails in question to be correctly regenerated; so it's currently my theory that the cut+paste, or more likely the removal and addition of files, causes SpaceFM to enter some kind of global state that prevents it from deciding that these files need a thumbnail. I say this partly because navigating to other directories has no effect (some files will be shown with no thumbnail, though they are ordinary JPG or PNG files).

[1] The Paste action in question is a custom action that is fairly simple -- it uses tmsu-fs-mv to move the file and update the TMSU database accordingly, and also schedules an update of an index file if one isn't already pending.

**Comment by IgnorantGuru** [comment]
I can't reproduce this with a normal Cut and Paste.  Perhaps something tmsu is doing is causing the issue.  You might try triggering a refresh after the copy:

    spacefm -s run-task refresh

Also, thumbnail loading has changed in the upcoming alpha code #582.  I doubt that will make a difference in this case, but you should note the output of 'spacefm --version' here.

In general, when you copy or move a file into the directory, SpaceFM is notified by inotify that the file has been created, and it will generate a thumbnail.  However, if being copied the file may be incomplete at that point.  Each time the file is written to as its being copied, SpaceFM receives a file-changed notification (regardless of what process changed the file) and queues an update of the file, which regenerates the thumbnail.  So when the last write is completed and the file date is changed, the last update to the thumbnail normally occurs.

**Comment by IgnorantGuru** [comment]
Instead of tmsu-fs-mv, note that you can also use

    mv path/to/file new/path && tmsu repair --manual path/to/file new/path

which may change the situation.

**Comment by 0ion9** [comment]
> I can't reproduce this with a normal Cut and Paste

Neither can I, consistently. It only happens sometimes -- which I really should have mentioned, sorry!
It's quite unclear what causes it.. it can have quite a long run of behaving normally, and then it seems to get into this mysterious state, after which only restarting appears to normalize things.

Version is currently 20150919.1537 ; I have encountered this behaviour over several different versions (I'm on Arch Linux x86_64, using the spacefm-git AUR package). Due to your question, I've just installed version 20151018, but have yet to test it.

> spacefm -s run-task refresh

Will do, however note that refreshing normally via F5 / View->Refresh has no effect on thumbnails (once they are in this state). That is one of the reasons contributing to my idea that something global gets in a weird state.

> mv path/to/file new/path && tmsu repair --manual path/to/file new/path

Would be pretty weird if it did change the situation, since that's pretty exactly what tmsu-fs-mv does, just with the safeguards removed:

```
#!/bin/sh

usage="\
Moves a file in the filesystem and updates the path in TMSU

Usage: $(basename $0) SRCFILE DESTFILE"

if test $# -ne 2; then
    echo "${usage}" 1>&2
    exit 1
fi

DEST=$2
if [ -d "$2" ]; then
    BASE=$(basename "$1")
    DEST="$DEST/$BASE"
fi

mv "$1" "$DEST" && tmsu repair --manual "$1" "$DEST"
```

BTW, I've pastebinned the current Paste script I'm using [here](https://bpaste.net/show/fd8b5200b4c1). The most opaque part of it is the references to 'update_refcodes', which is a personal script that schedules (using `at`) the update of the index where applicable (ie. if the source or target dir are within certain directory trees). Since it neither examines nor updates files directly, I considered it unlikely as a cause.

**Comment by 0ion9** [comment]
After more testing, I'm concluding it probably is related to #508 , as it appears to be 'okay' to create files in directories that aren't being viewed (eg. via GIMP) -- later viewing them, their thumbnail generates okay, as contrasted with creating files in directories that ARE currently viewe.

 I would like to test with a wider range of software, but currently it seems that either Paste action or GIMP will cause it -- the former occasionally, the latter consistently. So I have updated the summary accordingly.

I'm planning to try MTPaint, GrafX2, Inkscape, Tiled; I doubt GIMP does anything particularly special when saving files, but if there's an app that you think would be particularly informative I'm happy to try it.

**Comment by IgnorantGuru** [comment]
Changes in #508 introduced in 1.0.4 should only affect video thumbnails, not image thumbnails.

I tried saving an image in GIMP but still cannot reproduce this behavior.

SpaceFM places a watch on the directory using inotify (part of the kernel).  When a file is created or changes (is modified, size or date or permissions change, etc), SpaceFM is notified.  It adds the file to a changed/created queue which is processed at intervals (about every 500ms).  All the properties of the files in the queue are then reloaded and redisplayed, including the thumbnail.  Thumbnails store the mod date/time of the file they were created from, and if this date mismatches the current file mod date, the thumbnail is recreated.  (In the case of video files, 1.0.4 and later will only recreate the thumbnail if the mod time of the video file is greater than 5 seconds ago, per #508).

So overall the behavior is based on file modification times, internal thumbnail "tEXt::Thumb::MTime", and inotify events for created/changed files.  You can see the thumbnail load logic [here](https://github.com/IgnorantGuru/spacefm/blob/99b9d5d3269a6a7f42e8b39ec432901e2e4e6b1b/src/vfs/vfs-thumbnail-loader.c#L468).

For testing, I recommend the alpha branch (#582), because the other code is now somewhat obsolete (even if you find a problem, unless it exists in alpha also, it won't be fixed but merely replaced with alpha).

Thanks for the testing, yet thus far I don't see what could be causing what you describe.  The basic question is why is your system showing this problem when presumably other systems aren't?

**Comment by 0ion9** [comment]
The alpha install is now done. So far, I haven't reproduced the bug on it.

The most likely difference between my system and other (non-Arch, at least) systems would be the kernel version: 4.2.5
But just for the sake of completeness:

WM = QTile (python3 version, git 2324.5ca9be1)
Filesystem = Btrfs
Distro = Arch Linux x86_64

**Comment by IgnorantGuru** [comment]
> It initially may appear that thumbnails are not being generated at all, but when mousing over them, the thumbnail immediately appears.

Yeah I'm aware of that one.

Your other problems may be related to btrfs, as it tends to have bugs that affect issues like this (it just comes up often in issues).  Also, I don't believe the relevant code in SpaceFM has changed in a long time (years?), so if there was a problem in the code, it would probably be more generally evident by now.  You might see if another filesystem gives the same results.

It could be that when the file is first created, the notification is processed and the thumbnail generated before the file contents are complete (which is normal).  However, the final write and closure should trigger another update.  Perhaps for some reason that isn't occurring.

I'll leave this open to collect any further info.

**Comment by OmegaPhil** [comment]
I can confirm the problem where you have a huge GIMP project (e.g. 494MB in RAM, 4700x4416), exporting to PNG which takes some time - SpaceFM refreshes the thumbnail a few times, and sometimes the final thumbnail clearly has the bottom part missing - no amount of refreshing or moving directories changes things, presumably the generated thumbnail is just wrong.

I can queue this up to look at, although as usual I won't have a clue what to do about it to start with.

**Comment by OmegaPhil** [comment]
I use btrfs too for reference.

**Comment by IgnorantGuru** [comment]
@OmegaPhil Sure if you want to look at this since you can reproduce it...  Sounds complicated but it's not too bad: when a file changes, the dir monitor code leads [here](https://github.com/IgnorantGuru/spacefm/blob/99b9d5d3269a6a7f42e8b39ec432901e2e4e6b1b/src/vfs/vfs-dir.c#L441), which adds the file to the queue (a GList).  That queue is processed periodically [here](https://github.com/IgnorantGuru/spacefm/blob/99b9d5d3269a6a7f42e8b39ec432901e2e4e6b1b/src/vfs/vfs-dir.c#L1125).  That in turns runs [update_changed_files](https://github.com/IgnorantGuru/spacefm/blob/99b9d5d3269a6a7f42e8b39ec432901e2e4e6b1b/src/vfs/vfs-dir.c#L1050), which emits FILE_CHANGED_SIGNAL for the vfsdir (a signal which both the file browser and file list have handlers for), and it resets and reloads the info for the file via [update_file_info](https://github.com/IgnorantGuru/spacefm/blob/99b9d5d3269a6a7f42e8b39ec432901e2e4e6b1b/src/vfs/vfs-dir.c#L1005).  That resets the icon and triggers a reload of the thumbnail.

So inotify signals the changed file, which is handled by the dir monitor, which run a callback in vfs-dir.c, which eventually updates the info for the file ref and emits a signal on the vfsdir, which is handled by all file lists and file browsers that are using that vfsdir.  Also, the thumbnail loading is backgrounded, so when it completes creation, that thread emits another signal for the file list to update display.

My guess is that with btrfs, the final write call on the file triggers the file changed signal in the dir monitor.  This perhaps gets processed by SpaceFM, and the thumbnail recreated, before the data is finished writing to the file.  You can verify this by enabling some of the printfs I have in vfs-dir.c, so you can see the signals as the file is written.

Odd that we don't see this with slow network filesystems, though, or perhaps I never noticed it.  Or it could be only btrfs triggers this problem because of when the changed signals are emitted.

I designed the file changed/created/deleted queues (was inefficient without it) so I know that code well.  Unless there is some final changed signal when the last write completes, which there doesn't seem to be based on what you're reporting, then I think the only solution would be to add a delay between file changed signal and reload.  But it will be tricky where that delay is added and how, because it's already part of the queue mechanism (which already introduces a delay from 0ms to up to 200ms by collecting the signals in the queue).

Or, it could be nothing should be done for this.  Leave it slightly broken if it's a hack just for btrfs due to some problem btrfs has.  Obviously best if it can work better, but this may get messy.  I don't particularly like slowing things down with delays just for btrfs, if that's what it comes to.  Or maybe a smart hack could be added, eg vfs_dir_new already calls vfs_volume_dir_avoid_changes, which could store the filesystem type in the vfsdir struct, which could be used for a special hack that just applies for btrfs.  Efficient enough but hacky.

Overall, this is one of those issues that arises in SpaceFM because it has it's own vfs and it rides right on the kernel (as opposed to using gvfs, etc).  Some filesystems are less responsive or have different timing, so SpaceFM becomes less responsive, just as with network filesystems.

**Comment by IgnorantGuru** [comment]
Also, perhaps we should consider a more broad solution for this which also addresses #508 fully, since they are similar (thumbnails of a file with write in progress being updated properly).  The current solution for #508 is hackish and incomplete.

**Comment by IgnorantGuru** [comment]
Here's why I think this may be a bug in btrfs:  The purpose of the inotify watch is for a process to be signaled when a file is changed, so it can be reread, etc.  However, btrfs seems to emit the signal then change the file further.  This may be because it (or the kernel) triggers the signal when the write call happens, but due to its internal compression or whatever, the file is actually updated in the data cache sometime after that.  I think that's buggy behavior, and why only btrfs may be exhibiting this problem.

It's probably a race of a few ms.

**Comment by OmegaPhil** [comment]
I've just demonstrated this on an ext4 filesystem on an SSD, so its not really a btrfs thing (I'll keep testing in this environment).

**Comment by IgnorantGuru** [comment]
It's probably a similar timing issue between the last file change detection and when the content is actually readable, for whatever reason.

We probably need a mechanism which updates a thumbnail one last time a second or two after it's last changed (also for #508).  But you can imagine how inefficient that would be.  Or it should somehow delay processing, but with the FIFO queue that will be tricky, and it will slow down the application's responsiveness.  

Perhaps it could flag the file for a later update check, and only do the update if the file size has changed.  But that would have to be worked into the threading, including the fact that the file may be gone or the tab closed before it's processed, and it would require more memory, at least the 64bit file size (times thousands of files).

Overall I would say this requires some thumbnail loader design work (places to review being vfs-dir.c and vfs-thumbnail-loader.c).  It's probably not just a one line bug fix.  Somehow that delay would have to be accommodated.

**Comment by IgnorantGuru** [comment]
And fwiw I still think something is wrong with the filesystem(s) behavior:  Emitting a changed signal and then making further changes to the file size with no additional signal?  Sounds broken.  Defeats the whole purpose of the signal.

**Comment by IgnorantGuru** [comment]
One way to test this theory is to add a sleep right before the thumbnail loader creates the thumbnail.  Just have it sleep for 100ms or 500ms.  It's a background thread.

Would it hurt for it to wait 100ms before it loads every thumbnail?  Probably not appreciably, but it seems inefficient.  For every 10 thumbnails being loaded that takes 1 second longer.  It would be better if it only added the delay when needed (when the thumbnail hasn't already been sitting in the loader queue due to a backlog).  Maybe a precision timestamp could be added to the thumbnail loader request struct?  When the request is processed, if the timestamp is too recent, it could add a delay.  That might work.

**Comment by OmegaPhil** [comment]
Hmm, with the latest code, 10 GIMP exports of a large PNG all work fine. Will continue to fiddle after dinner.

**Comment by OmegaPhil** [comment]
OK, with a harder example 6/20 (30%) extractions fail, but with 500ms delay they all pass. Shall I research into the 100ms timestamp option?

**Comment by IgnorantGuru** [comment]
Yeah I think if you add a conditional delay in [this else block](https://github.com/IgnorantGuru/spacefm/blob/69fa2e73f0035c9fb68e14dbb4cacd52b0e8b584/src/vfs/vfs-thumbnail-loader.c#L301), that might work.  The timestamp can go in [this ThumbnailRequest struct](https://github.com/IgnorantGuru/spacefm/blob/69fa2e73f0035c9fb68e14dbb4cacd52b0e8b584/src/vfs/vfs-thumbnail-loader.c#L57).  So every time it adds a thumbnail request it includes the timestamp of when it was added.

If it tries to process them thumbnail before time - timestamp < delay_minimum, then it should add the delay.  That way ones that are sitting in the queue with a backlog won't need the delay.

The trick will be to use a precision timestamp, as seconds are ofc too coarse.  I think there's a time function that gives milliseconds since epoch, etc.  You might find it used elsewhere in the source.  And maybe g_usleep for the sleep.

The thumbnail loader already deals with eg tabs being closed while requests are queued (file are ref counted), so that should be a good place to put it.

Just remember your system is fast, so maybe add 100ms more or so to whatever you find to be a reasonable minimum on your fast system.

**Comment by IgnorantGuru** [comment]
Another thing to keep in mind: it doesn't really need the delay unless it's really going to create a new thumbnail.  If it skips it or reloads the existing thumbnail in the load function, then no delay is needed.  So to not slow things down unnecessarily, the conditional delay should go in the load function itself, after all the other conditions for creating the thumbnail have been satisfied.

**Comment by OmegaPhil** [comment]
Yeah I thought that was too high in the stack. I was looking at vfs-thumbnail-loader.c:~600, just after if '( file_is_video == FALSE )', before the gdk_pixbuf_new_from_file_at_size call.

**Comment by IgnorantGuru** [comment]
Sounds good.

**Comment by IgnorantGuru** [comment]
Also for the initial dir load, no delays, so the request function should have a gboolean option for delay, only used when called from vfs-dir.c on file changed/created.

And noting here the GTimer - if you add a GTimer to the req struct and start it, you can probably just check elapsed time, rather than using timestamps.

**Comment by OmegaPhil** [comment]
Just to confirm, it doesn't look like vfs-dir.c has anything to do with thumbnails outside of desktop file icons - it looks like I should be fiddling with vfs-thumbnail_loader_request, which is (ignoring the desktop implementation) called from various places in ptk-file-list.c?

**Comment by IgnorantGuru** [comment]
Yeah vfs-dir.c contains the code that receives the 'file changed' signal for the directory.  It queues that signal and then processes it, triggering a reload of the file properties.  It then emits a 'file-changed' signal handled in ptk-file-list.c, which requests the thumbnail.

**Comment by OmegaPhil** [comment]
See above ^, 400ms is enough delay for me, but have doubled as requested. I've implemented it such that a timer is only created when there is a need to delay the load.

**Comment by IgnorantGuru** [comment]
Cursory review looks good, thanks.  I suggest we start with 600ms and see how that performs generally.  Also, it seems that the g_usleep should be for 600ms minus elapsed, rather than a fixed value?  Iow count the time spent in the queue, if any, toward the needed sleep time to achieve minimum required delay.

@0ion9 If you can help us test this version, it's the alpha branch.  You can install it with the net installer like this:

    spacefm-installer --version=alpha

**Comment by IgnorantGuru** [comment]
Also you might verify that the delay is actually needed for the file_created signal.  Normally a write and file_changed signal will always follow a creation (unless it's a same-filesystem move operation, in which case delay isn't needed anyway).  So you may be doubling the delay by doing it for both created and changed.

**Comment by IgnorantGuru** [comment]
@OmegaPhil Please note that I just pushed a fix for #508 to alpha (a28fc3ac), so pull before working.  I didn't change any of your code, I just added to some of your comments.  Most of my changes were in vfs-dir.c.  Also your testing of that fix is welcome.

**Comment by 0ion9** [comment]
@IgnorantGuru works for me so far, however I don't have time to do careful testing atm. In three days I'll report back.

**Comment by IgnorantGuru** [comment]
@OmegaPhil Actually, I think I'd like to consolidate these and use the [solution for #508](https://github.com/IgnorantGuru/spacefm/commit/a9ecdea0c3e68e550a0c2c858603a624c46271b2) for images too if it works for this issue, which I think it will.  I have reduced the delay to 1 second for both images and videos.  I pushed that to the next branch so you can test it there, and alpha still contains your fix.

I think most images will show the thumbnail immediately on the create signal anyway, but for those that show this bug, the update 1 second later should correct it.

What it does is it now keeps a separate file-change queue for changing image/video files if thumbnails are enabled. This queue is processed after 1.0-1.5 seconds providing that no image/video files in that dir change in that period, otherwise the timer is reset.

So when copying or downloading a file, you should see the file size updating at the normal rate, but the thumbnail will only be loaded when the file is initially created (if possible, otherwise icon shown). Then it won't be updated again until about 1 second after the last write to the file. Or anytime there is a file writing pause greater than 1 second, it will be updated then also.

**Comment by IgnorantGuru** [comment]
Also in next, I have made it flush the delayed queue every 7 seconds or so.  Otherwise when downloading or copying multiple files, no thumbnails would be shown until they all completed.  This way it will attempt to load the thumbnail every few seconds anyway.

**Comment by OmegaPhil** [comment]
next passes the 20 extractions test on ext4, btrfs, and in most cases on aufs + btrfs - I've had 2/20 failures in a directory that had a downloading video that had since stopped writing (the failures occurred long after the writing stopped). I haven't been able to recreate the failure, so I'm assuming its just something dodgy about aufs.

**Comment by IgnorantGuru** [comment]
Okay thanks for testing.  Let me know if you reproduce that.  I would think if several files are downloading/writing, and one stops, the others would continue to update.  In fact we can simulate this by pausing a copy task in the task manager.  But if the filesystem became unresponsive at a system call, this may have halted that thread.

I'm going to close this issue but anyone feel free to comment further with any results.  Also, I'm deleting the alpha branch from origin, but I'll keep a local copy for awhile so we have the other code.  Please use the next branch to test this fix, due in 1.0.5 release.

**Comment by IgnorantGuru** [comment]
Please note that the fix for this is once again in the alpha branch, not next, [as explained](https://github.com/IgnorantGuru/spacefm/issues/582#issuecomment-159708250).
